kvtests/common_tests.go

```diff
@@ -0,0 +1,539 @@
+var bytesValue = bytesKey.Next().Bytes()
+var largerBytesKey = stoabs.BytesKey([]byte{4, 5, 6})
+var largerBytesValue = []byte{100, 101, 102}
+
+const stringKey = "message"
+const stringValue = "Hello, World!"
+
+const shelf = "test"
+
+type StoreProvider func(t *testing.T) (stoabs.KVStore, error)
+
+func TestReadingAndWriting(t *testing.T, storeProvider StoreProvider) {
+	t.Run("read/write", func(t *testing.T) {
+		t.Run("write, then read ([]byte])", func(t *testing.T) {
+			store := createStore(t, storeProvider)
+
+			err := store.Write(func(tx stoabs.WriteTx) error {
+				writer, err := tx.GetShelfWriter(shelf)
+				if err != nil {
+					return err
+				}
+				return writer.Put(bytesKey, bytesValue)
+			})
+			if !assert.NoError(t, err) {
+				return
+			}
+
+			var actual []byte
+			err = store.ReadShelf(shelf, func(reader stoabs.Reader) error {
+				actual, err = reader.Get(bytesKey)
+				return err
+			})
+			assert.NoError(t, err)
+			assert.Equal(t, bytesValue, actual)
+		})
+		t.Run("write, then read (string)", func(t *testing.T) {
+			store := createStore(t, storeProvider)
+
+			err := store.Write(func(tx stoabs.WriteTx) error {
+				writer, err := tx.GetShelfWriter(shelf)
+				if err != nil {
+					return err
+				}
+				return writer.Put(stoabs.BytesKey(stringKey), []byte(stringValue))
+			})
+			if !assert.NoError(t, err) {
+				return
+			}
+
+			var actual []byte
+			err = store.ReadShelf(shelf, func(reader stoabs.Reader) error {
+				actual, err = reader.Get(stoabs.BytesKey(stringKey))
+				return err
+			})
+			assert.NoError(t, err)
+			assert.Equal(t, stringValue, string(actual))
+		})
+
+		t.Run("ReadShelf for non-existing shelf", func(t *testing.T) {
+			store := createStore(t, storeProvider)
+
+			called := false
+			err := store.ReadShelf(shelf, func(reader stoabs.Reader) error {
+				called = true
+				return nil
+			})
+
+			assert.NoError(t, err)
+			assert.False(t, called)
+		})
+
+		t.Run("GetShelfReader for non-existing shelf", func(t *testing.T) {
+			store := createStore(t, storeProvider)
+
+			err := store.Read(func(tx stoabs.ReadTx) error {
+				bucket, err := tx.GetShelfReader(shelf)
+				if err != nil {
+					return err
+				}
+				if bucket == nil {
+					return nil
+				}
+				t.Fatal()
+				return nil
+			})
+			assert.NoError(t, err)
+		})
+
+		t.Run("read non-existing key", func(t *testing.T) {
+			store := createStore(t, storeProvider)
+
+			err := store.WriteShelf(shelf, func(writer stoabs.Writer) error {
+				return writer.Put(stoabs.BytesKey(stringKey), []byte(stringValue))
+			})
+			if !assert.NoError(t, err) {
+				return
+			}
+
+			var actual []byte
+			err = store.ReadShelf(shelf, func(reader stoabs.Reader) error {
+				actual, err = reader.Get(bytesKey)
+				return err
+			})
+			if !assert.NoError(t, err) {
+				return
+			}
+			assert.Nil(t, actual)
+		})
+	})
+}
+
+func TestRange(t *testing.T, storeProvider StoreProvider) {
+	t.Run("range", func(t *testing.T) {
```

Comment:
add a test with a string key type and check the ordering returned.
KEY_0 -> KEY_1 -> KEY_2 must be returned in the correct order.

Reply:
Newly added tests assert this.